python/cuml/cuml/accel/estimator_proxy.py

```diff
@@ -12,122 +12,325 @@
+        # The GPU estimator. This is only non-None if an estimator was successfully fit
+        # on the GPU.
+        self._gpu = None
+        # Whether fit attributes on the CPU estimator are currently in sync with those
+        # on the GPU estimator.
+        self._synced = False
+
+    def __init_subclass__(cls, **kwargs: Any) -> None:
         super().__init_subclass__(**kwargs)
-        cls.import_cpu_model()
-        cls._cpu_model_sig = inspect.signature(cls._cpu_model_class)
-        cls._proxy_module = cls.__module__
-        cls.__module__ = cls._cpu_model_class.__module__
-        cls.__qualname__ = cls._cpu_model_class.__qualname__
-        cls.__doc__ = cls._cpu_model_class.__doc__
-        cls.__init__.__signature__ = cls._cpu_model_sig
-
-    def __init__(self, *args, **kwargs):
-        # The cuml signature may not align with the CPU signature.
-        # Additionally, some models support positional arguments.
-        # To work around this, we
-        # - Bind arguments to the CPU signature
-        # - Convert the arguments to named parameters
-        # - Translate them to cuml equivalents
-        # - Then forward them on to the cuml class
-        bound = self._cpu_model_sig.bind_partial(*args, **kwargs)
-        translated_kwargs, self._gpuaccel = self._hyperparam_translator(
-            **bound.arguments
-        )
-        super().__init__(**translated_kwargs)
-        self.build_cpu_model(**kwargs)
-
-    def __repr__(self):
-        return self._cpu_model.__repr__()
-
-    def __str__(self):
-        return self._cpu_model.__str__()
-
-    def __getattr__(self, attr):
-        # Don't dispatch __sklearn_clone__ so that cloning works as a
-        # as a regular estimator without __sklearn_clone__
-        may_dispatch = attr != "__sklearn_clone__"
-
-        if may_dispatch and hasattr(self._cpu_model_class, attr):
-            self.build_cpu_model()
-            self.gpu_to_cpu()
-            return getattr(self._cpu_model, attr)
-        raise AttributeError(
-            f"{type(self).__name__!r} object has no attribute {attr!r}"
-        )
+
+        # Forward some metadata to make the proxy class look like the CPU class
+        cls.__qualname__ = cls._cpu_class.__qualname__
+        cls.__doc__ = cls._cpu_class.__doc__
+        cls.__signature__ = inspect.signature(cls._cpu_class)
+        # We use the `_cpu_class_path` (which matches the patched path) instead of
+        # the original full module path so that the class (not an instance) can
+        # be pickled properly.
+        cls.__module__ = cls._gpu_class._cpu_class_path.rsplit(".", 1)[0]
+
+        # Add proxy method definitions for all public methods on CPU class
+        # that aren't already defined on the proxy class
+        methods = [
+            name
+            for name in dir(cls._cpu_class)
+            if not name.startswith("_")
+            and callable(getattr(cls._cpu_class, name))
+        ]
```

Comment:
What do you think about using `inspect.getmembers()` combined with `inspect.ismethod()` to do this instead? I've not looked at how those are implemented, just assuming that if the `inspect` module maintainer made the effort to define these helpers that there are traps to fall into which we can avoid by using `inspect`

A potential trickyness is that naively `getmembers()` needs an instance, not a class :-/

Reply:
I mostly didn't know about those `inspect` methods. I do want something simple that works with a class and not an instance though, will poke around `inspect` to see if there's something better we should be using, but I don't think what is here is that bad.